doc/developer/design/20220214_create_cluster.md

```diff
@@ -0,0 +1,248 @@
+from the active cluster. In the future, we may add optional `IN CLUSTER` clauses
+to these statements, but for simplicitly, they are omitted for now; reliably
+extending the `SELECT` grammar is difficult.
+
+### System catalog changes
+
+A new `mz_clusters` table with the following structure will describe the
+available clusters in the system:
+
+Field   | Type      | Meaning
+--------|-----------|--------
+id      | `bigint`  | The ID of the cluster.
+name    | `text`    | The name of the cluster.
+virtual | `boolean` | Whether the cluster is virtual.
+size    | `text`    | The size of the cluster. `NULL` if the cluster is virtual.
+
+The existing `mz_indexes` table will be extended with a `cluster_id` field:
+
+Field      | Type      | Meaning
+-----------|-----------|--------
+...        | ...       | ...
+cluster_id | `bigint`  | The ID of the cluster in which the index is materialized.
+
+The existing `mz_sinks` table will be extended with a `cluster_id` field:
+
+Field      | Type      | Meaning
+-----------|-----------|--------
+...        | ...       | ...
+cluster_id | `bigint`  | The ID of the cluster in which the sink is maintained.
+
+### Coordinator changes
+
+Internally, each cluster will be identified by an identifier in the range `[1,
+i64::MAX]`. The coordinator will be responsible for allocating identifiers and
+durably recording the mapping from names to identifiers via the following new
+table in the SQLite catalog:
+
+```sql
+CREATE TABLE clusters (
+    id   integer PRIMARY KEY,
+    name text NOT NULL UNIQUE
+);
+```
+
+This table works analogously to the existing `databases` table in the catalog.
+Initially, all clusters in the table will be virtual. Storing the `SIZE` option
+of a non-virtual cluster is deferred to future work.
+
+The migration that creates the table will additionally install a default row
+of `(1, 'default')`. This creates a cluster named `default` that is present by
+default in every Materialize installation. The default cluster is not otherwise
+special and can be dropped if desired.
+
+We will write a catalog migration that adds the `IN CLUSTER default` clause to
+all existing indexes.
+
+The various bookkeeping in the coordinator will need to change to account for
+clusters. Execution of a `SELECT` statement, for example, will need to select
+from only the indexes in the active cluster.
+
+TODO(mcsherry?): can we more specifically describe what bookkeeping will change?
+
+### `mz_cluster_id`
+
+The existing `mz_cluster_id` function uses "cluster" to mean something entirely
+unrelated to the new concept of clusters. To reduce confusion, we need to
+introduce a breaking change to rename this function ASAP. I propose
+`mz_legacy_installation_id` as a strawman.
+
+In the future, I believe we'll instead want a function like `mz_account_id` that
+corresponds to the [account](../platform/ux.md#account) ID in Materialize Platform.
+
+## Future work
```

Comment:
Do we envision only exposing clusters' underlying machine's resource usage/capacity through some dashboard?

Reply:
Is this an oblique way of suggesting a future work item like the following?

> * Introduce logging sources in the system catalog like that show the active cluster's resource usage (e.g., `mz_cluster_cpu_usage` and `mz_cluster_memory_usage`). 

Because I think that makes a lot of sense! If we do have such a dashboard, I think we might want it to pull metrics from exactly that sort of logging view.

Reply:
Yeah, exactly